File: coding/src/com/algos/core/li20_math/revised/L66_PlusOne.java
package  com.algos.core.li20_math.revised;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

public class L66_PlusOne {

	public static void main(String[] args) {
		L66_PlusOne obj = new L66_PlusOne();
		System.out.println(Arrays.toString(obj.plusOne(new int[] {1,2,3})));
	}

	// Problem: https://leetcode.com/problems/plus-one/
    // Idea: In 999 case carry could propagate from right to left. Hence, process right to left.
	public int[] plusOne(int[] digits) {
        List<Integer> list = new ArrayList<>();
        int carry = 1, n=digits.length;
        for(int i=n-1; i>=0; i--){
            int sum = carry+digits[i];
            list.add(sum%10);
            carry = sum/10;
        }
        if(carry > 0)
            list.add(carry);
        
        Collections.reverse(list);
        return list.stream().mapToInt(i->i).toArray();
    }
}

File: dsa/topics/20_math/python/l66_plus_one.py
# Problem: https://leetcode.com/problems/plus-one/
# Idea: In 999 case carry could propagate from right to left. Hence, process right to left.
def plus_one(digits):
    result = []
    carry = 1
    n = len(digits)

    for i in range(n - 1, -1, -1):
        s = carry + digits[i]
        result.append(s % 10)
        carry = s // 10

    if carry > 0:
        result.append(carry)

    result.reverse()
    return result


# ------------------ Test (Same as Java main) ------------------

print(plus_one([1, 2, 3]))  # Expected: [1, 2, 4]
